File: Coding_1124 Backtracking/이강훈/이강훈/2번.java
/*
https://school.programmers.co.kr/learn/courses/30/lessons/172927

1. 아이디어
백트래킹 구현 (사용 곡갱이 수, 사용 미네랄 수, 비용)

2. 자료구조
배열

3. 시간 복잡도
O(N * M) // N = 곡갱이 수, M = 미네랄 수
*/

import java.util.Arrays;

public class Solution {
    int[] picks;
    String[] minerals;
    int answer = Integer.MAX_VALUE;
    int totalPick;
    int[][] costByPick = new int[][]{{1, 1, 1}, {5, 1, 1}, {25, 5, 1}};

    public int solution(int[] picks, String[] minerals) {
        this.picks = picks;
        this.minerals = minerals;
        this.totalPick = Arrays.stream(picks).sum();

        backtrack(0, 0, 0);

        return answer;
    }

    public void backtrack(int usedPick, int usedMineral, int cost) {
        if (usedPick == totalPick || usedMineral == minerals.length) {
            answer = Math.min(answer, cost);

            return;
        }

        for (int i = 0; i < 3; i++) {
            if (picks[i] > 0) {
                int currentCost = 0;

                int j = usedMineral;
                for (; j < Math.min(usedMineral + 5, minerals.length); j++) {
                    if (minerals[j].equals("diamond")) currentCost += costByPick[i][0];
                    else if (minerals[j].equals("iron")) currentCost += costByPick[i][1];
                    else if (minerals[j].equals("stone")) currentCost += costByPick[i][2];
                }

                picks[i]--;
                backtrack(usedPick + 1, j, cost + currentCost);
                picks[i]++;
            }
        }
    }
}

File: Coding_1124 Backtracking/이강훈/최원준/2번.py
# https://school.programmers.co.kr/learn/courses/30/lessons/172927

'''
1. 아이디어 :
    백트래킹으로 풀 수 있다.
    5개씩 빼므로, 처음 미네랄을 5배수로 만든다.
    calc_fatigue 메서드를 통해서 피로도를 계산한다.
    temp 배열을 만들어서, minerals에서 꺼낸 것들을 기록한다.
2. 시간복잡도 :
    O( len(picks) ** math.ceil(minerals%5) ) + O(mineral log mineral)
3. 자료구조 :
    배열
'''


def solution(picks, minerals):
    def backtrack(tired):
        if not minerals or sum(picks) == 0:
            ans[0] = min(ans[0], tired)
            return

        for i in range(len(picks)):
            if picks[i] != 0:
                temp = []

                picks[i] -= 1
                for _ in range(5):
                    m = minerals.pop()
                    tired += calc_fatigue(i, m)
                    temp.append(m)

                backtrack(tired)

                picks[i] += 1
                for _ in range(5):
                    m = temp.pop()
                    tired -= calc_fatigue(i, m)
                    minerals.append(m)

    def calc_fatigue(pick_idx, mineral):
        if mineral == "stone":
            return 1
        if mineral == "iron":
            return 5 if pick_idx == 2 else 1
        if mineral == "diamond":
            return 25 if pick_idx == 2 else 5 if pick_idx == 1 else 1
        return 0 #dummy

    while len(minerals) % 5 != 0:
        minerals.append("dummy")
    minerals = minerals[::-1]

    ans = [float('inf')]

    backtrack(0)
    return ans[0]
